### PR TITLE
feat(dev): add Cypress and Puppeteer to GitPod

### LINKED ISSUE #39728: GitPod Environment Not Set Up for Tests
When running an instance of the platform in GitPod, `npm run test:curriculum` throws this error:

```console
> cross-env FULL_OUTPUT=true mocha --delay --reporter progress

Error: Failed to launch the browser process!
/workspace/freeCodeCamp/curriculum/node_modules/puppeteer/.local-chromium/linux-800071/chrome-linux/chrome: error while loading shared libraries: libnss3.so: cannot open shared object file: No such file or directory


TROUBLESHOOTING: https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md

    at onClose (/workspace/freeCodeCamp/curriculum/node_modules/puppeteer/lib/cjs/puppeteer/node/BrowserRunner.js:193:20)
    at Interface.<anonymous> (/workspace/freeCodeCamp/curriculum/node_modules/puppeteer/lib/cjs/puppeteer/node/BrowserRunner.js:183:68)
    at Interface.emit (events.js:327:22)
    at Interface.close (readline.js:416:8)
    at Socket.onend (readline.js:194:10)
    at Socket.emit (events.js:327:22)
    at endReadableNT (_stream_readable.js:1220:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @freecodecamp/curriculum@0.0.0-next.4 test:full-output: `cross-env FULL_OUTPUT=true mocha --delay --reporter progress`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @freecodecamp/curriculum@0.0.0-next.4 test:full-output script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/gitpod/.npm/_logs/2020-09-29T19_02_52_374Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @freecodecamp/freecodecamp@0.0.1 test-curriculum-full-output: `cd ./curriculum && npm run test:full-output`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @freecodecamp/freecodecamp@0.0.1 test-curriculum-full-output script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/gitpod/.npm/_logs/2020-09-29T19_02_52_393Z-debug.log
gitpod /workspace/freeCodeCamp $ 
```

Since the update of Puppeteer, I needed to run `sudo apt install libgbm1` to update, before the tests work. I assume something similar can be done for GitPod.

Useful resources courtesy of Mrugesh: 

Our config:
https://github.com/freeCodeCamp/freeCodeCamp/blob/master/.gitpod.yml

Documentation:
https://www.gitpod.io/docs/config-docker/
https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md#running-on-alpine

### PR DESCRIPTION
**Pull request by ShaunSHamilton**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

(01/02/2021) - requires to be on the Feature Preview of GitPod

1) Run `npm run cypress:install-build-tools`
2) Ensure to select the correct keyboard layout, when prompted during the installation
3) Run `cypress:dev:run` with `server` and `client` running in different terminals (background)

A DOCS PR is soon to follow

Closes #39728 

P.S. _I expect some tests to fail, because some are hard-coded to work in a 'normal' dev environment. Specifically, values like 'localhost:8000' do not work in GitPod. So, another PR would need to take a look at that. I have done some of it in this PR (and a previous), which is why CYPRESS_BASE_URL is in the .gitpod config_

### CODE REVIEW

File: .gitpod.yml [hunk 1]
```diff
@@ -1,60 +1,62 @@
-image: gitpod/workspace-mongodb
+image:
+  file: .gitpod.Dockerfile
```
Comment (ShaunSHamilton) [inline .gitpod.yml:2]:
```suggestion
image: gitpod/workspace-mongodb
```

**Comment by ShaunSHamilton** [comment]
This is working as expected (on my end). To use, the workflow is:
1) Open new terminal in GitPod env: `npm run cypress:install-build-tools`
2) Be sure to select your keyboard layout, when prompted in the terminal
3) Use Cypress: `npm run cypress:dev:run`

The white-space changes appear to be in-line with our config, so I have left them. Let me know otherwise.

**Comment by raisedadead** [comment]
I ran the first step in the comment above: `npm run cypress:install-build-tools` and I see this.

image

**Comment by ShaunSHamilton** [comment]
@raisedadead Are you on the Feature Preview of GitPod? https://www.gitpod.io/docs/feature-preview/

**Review by raisedadead — APPROVED** [review]
I missed that in the conversation from the earlier PR. It works now, we may need to document this additional step.

LGTM. Will wait for @ojeytonwilliams to merge

**Comment by ahmadabdolsaheb** [comment]
Runs for me as well with steps from description.

Screen Shot 2021-02-04 at 6 46 40 PM

**Review by raisedadead — APPROVED** [review]
OK - I think here is what we should do.

1. I am going to land this in.
2. We add some docs about enabling the feature in settings, and running tests.
3. We work towards making the tests not rely on `localhost:8000`, but more generally compatible with a URL.

**Comment by ojeytonwilliams** [comment]
Hey Shaun, congrats on figuring this out.

So, I ran into

```
We detected that the Chromium Renderer process just crashed.

This is the equivalent to seeing the 'sad face' when Chrome dies.

This can happen for a number of different reasons:

- You wrote an endless loop and you must fix your own code
- There is a memory leak in Cypress (unlikely but possible)
- You are running Docker (there is an easy fix for this: see link below)
- You are running lots of tests on a memory intense application
- You are running in a memory starved VM environment
- There are problems with your GPU / GPU drivers
- There are browser bugs in Chromium

You can learn more including how to fix Docker here:

https://on.cypress.io/renderer-process-crashed
```

when I tried to `npm cypress:dev:run`, did anyone else see that?

Assuming it's not just me, then the issue is that Gitpod is running a Docker container, but we don't have control over it.  That means we may not be able to deploy the fixes shown here: https://github.com/cypress-io/cypress/issues/350 

Any thoughts on this?

**Comment by ShaunSHamilton** [comment]
> Any thoughts on this?

@ojeytonwilliams I have not seen that, this time around. Did you do anything different to the steps in the OP? (I am specifically asking about whether the issue is you did/did-not have the `client`/`server` up)

**Comment by raisedadead** [comment]
I am not quite sure about the messages above, but it's fairly possible I may have missed it because Shaun did say we can expect tests to fail.

**Comment by ojeytonwilliams** [comment]
So, I launched Gitpod via https://gitpod.io/#https://github.com/freeCodeCamp/freeCodeCamp, waited until the client and server were both running and then ran

```sh
npm run cypress:install-build-tools # picking 30 and 1, for the keyboard, though I doubt that matters!
npm run cypress:dev:run
```

and got the same error I mentioned above.  I was using Brave, not Chrome, so that might be a factor.  I'll try again and report back.

**Comment by ojeytonwilliams** [comment]
@ShaunSHamilton same error again I'm afraid.  It might be my machine that's the problem, but it's happening pretty consistently.

**Comment by ShaunSHamilton** [comment]
@ojeytonwilliams
 > It might be my machine that's the problem, but it's happening pretty consistently.

Thanks, for letting us know. I cannot see how it would be related to someone's PC. I would like to help debug this, later over chat.

**Comment by ojeytonwilliams** [comment]
For future reference, in case I forget... Shaun and I tested this some more and found that the problem occurs somewhat randomly.  The plan from here is to fix the port problem and then reevaluate if this annoying enough to try and fix.

**Pull request merged**